test/creds_test.go

```diff
@@ -545,3 +552,177 @@ func (s) TestServerCredsDispatch(t *testing.T) {
 		t.Errorf("Read() = %v, %v; want n>0, <nil>", n, err)
 	}
 }
+
+type clientCertificates struct {
+	validCert      *tls.Certificate
+	selfSignedCert *tls.Certificate
+	expiredCert    *tls.Certificate
+}
+
+func (c *clientCertificates) generateCertificates(ca *tls.Certificate) error {
+	var err error
+
+	c.validCert, err = c.generateCert(ca, time.Now().Add(time.Hour))
+	if err != nil {
+		return fmt.Errorf("failed to generate valid cert: %v", err)
```

Comment:
You should be able to simplify things significantly by just panicking whenever an error happens inside `generateCert`.  Failing "gracefully" for something we expect to pass, during initialization, isn't important.